### PR TITLE
Parquet support for Structs

### PR DESCRIPTION
**Pull request by razajafri**
This PR adds support for Structs in Parquet writer. It depends on https://github.com/rapidsai/cudf/pull/7998

### CODE REVIEW

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
CI is failing because of the cudf dependency

**Comment by revans2** [comment]
Moved to draft because what it depends on is not merged in so no point in worrying about CI right now.

File: docs/additional-functionality/cache-serializer.md
Comment (revans2) [inline docs/additional-functionality/cache-serializer.md:55]:
```
supported. If an unsupported type is encountered the Rapids Accelerator for Apache Spark will fall back to using the CPU for caching.
```

File: shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala
Comment (revans2) [inline shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala:281]: nit: I assume this is a follow on issue? If so it might be cleaner to say

```
// TODO: when arrays are supported for cudf writes add it here.
```

File: shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala
Comment (revans2) [inline shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala:457]: First off `cacheAttributes` is not used any more so it should be removed from the method signature. Second parquet has limits on the names it supports. It also does not want to have duplicate names, because reading it is ambiguous.  Spark does not have the same limitations. It is fine with duplicate names, because the expression id disambiguates them. Are we sure that this change will never run into limitations on parquet names?

File: shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala
Comment (revans2) [inline shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala:282]: Is there a follow on issue for this? Probably not that important. We could hack around it, but it is probably not worth it to do so.

File: shims/spark311/src/main/scala/org/apache/spark/sql/rapids/shims/spark311/GpuInMemoryTableScanExec.scala
Comment (revans2) [inline shims/spark311/src/main/scala/org/apache/spark/sql/rapids/shims/spark311/GpuInMemoryTableScanExec.scala:35]: A lot of this is white space changes, which makes it really hard to review. Could you revert the changes to just indentation please?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala:111]: We need a P1 blocker issue filed against us for this too.  At a minimum we need to fall back to the CPU for this.

File: shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala
Comment (razajafri) [reply shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala:282]: I have added it to this issue https://github.com/NVIDIA/spark-rapids/issues/1143

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala
Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala:111]: I created this when I saw the issue. https://github.com/rapidsai/cudf/issues/8070

Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala:111]: This is the plugin bug tracking this https://github.com/NVIDIA/spark-rapids/issues/2302

File: shims/spark311/src/main/scala/org/apache/spark/sql/rapids/shims/spark311/GpuInMemoryTableScanExec.scala
Comment (razajafri) [reply shims/spark311/src/main/scala/org/apache/spark/sql/rapids/shims/spark311/GpuInMemoryTableScanExec.scala:35]: This file only has indentation changes. I can revert and create a separate patch for it

Comment (revans2) [reply shims/spark311/src/main/scala/org/apache/spark/sql/rapids/shims/spark311/GpuInMemoryTableScanExec.scala:35]: I would prefer a separate patch.

File: shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala
Comment (razajafri) [reply shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala:458]: I didn't know about Spark not having a limitation on col names, thanks. The reason why I made the change was because now we are forcing people to set the column name on the column before writing parquet, but there's no guarantee there won't be duplicates. I will revert.

**Comment by razajafri** [comment]
@revans2 PTAL, I have added code to fallback to CPU if either `int96AsTimestamp` or `outputTimestamp == INT96` is true. Once I get an approval on this, I will merge the cudf change

**Comment by razajafri** [comment]
fixes #2302

**Comment by razajafri** [comment]
I made necessary changes to the tests so we avoid the overflow scenario when writing values to INT96.

**Review by revans2 — COMMENTED** [review]
We made the test avoid the INT96 overflow issue but we didn't fix the issue or document it in any way. If @sameerz really wants to get this merged in soon and do the INT96 work as a follow on, then I am fine with it. But I would much rather just fix it now, especially if the tests are passing now, even if we don't know 100% why.

File: integration_tests/src/main/python/parquet_write_test.py
Comment (razajafri) [reply integration_tests/src/main/python/parquet_write_test.py:40]: Spark doesn't overflow and writes the value as a binary. 

Thanks for pointing this out. I have made the change to throw an exception in case of overflow. PTAL

**Comment by razajafri** [comment]
@revans2 have I addressed your concerns? Please let me know

File: integration_tests/src/main/python/parquet_write_test.py
Comment (revans2) [reply integration_tests/src/main/python/parquet_write_test.py:40]: @razajafri we need a follow on issue to "fix" this eventually. It might not ever be a high enough priority to actually fix, but we should still have it on the backlog.

@jlowe what do you think? In a query we are never really writing more than one dataset per query so if they hit the error then can turn off parquet writes. Part of me wants a separate config, but another part says that is complexity for no real benefit.

File: docs/compatibility.md
Comment (revans2) [inline docs/compatibility.md:249]: Can we document how to work around this? like disable the parquet writer with the config ...

**Comment by razajafri** [comment]
> @razajafri we need a follow on issue to "fix" this eventually. It might not ever be a high enough priority to actually fix, but we should still have it on the backlog.

There is already an open issue tracking this https://github.com/rapidsai/cudf/issues/8070

**Comment by revans2** [comment]
> There is already an open issue tracking this [rapidsai/cudf#8070](https://github.com/rapidsai/cudf/issues/8070)
> 
> Once that is fixed we can remove the test boundaries

Can we have one on the plugin side? because we would need to remove the checks/docs that this is adding in. It would also be good to have traceability in the comments for both the tests and the code as well.

**Comment by razajafri** [comment]
> > There is already an open issue tracking this [rapidsai/cudf#8070](https://github.com/rapidsai/cudf/issues/8070)
> > Once that is fixed we can remove the test boundaries
> 
> Can we have one on the plugin side? because we would need to remove the checks/docs that this is adding in. It would also be good to have traceability in the comments for both the tests and the code as well.

I think you caught me while I was creating the plugin side issue. :)

File: docs/compatibility.md
Comment (razajafri) [inline docs/compatibility.md:253]: Let me change the wording of that sentence, that doesn't sound right

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala:72]: nit: we often use something like `s"setting the ${RapidsConf.ENABLE_PARQUET_INT96_WRITE.key} to true."` just to avoid any possible typos.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala:300]: bitCastTo returns a ColumnView that still has to be closed. You cannot chain them like this.

Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala]: Sorry, I keep forgetting the ColumnView is also a CPP side object and important to be closed

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala
Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala]: Fixed

**Review by revans2 — DISMISSED** [review]

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Review by jlowe — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Pull request merged**
